### DISCUSSION TITLE
How to configure "on device" commands?

### DISCUSSION THREAD

**Discussion by obones** [toverainc/willow#155] [Q&A] [answered]
Hello,

The readme file says this: 

> Configure up to 400 commands completely on device

I'm not sure if it's available just yet, but I believe it would be based on multinet which is the solution used in the espressif demo to get it to react to "turn on the light".
When I first played with the S3-BOX, I figured out from the sources and other discussions that it uses (used?) [g2p](https://github.com/Kyubyong/g2p) to get an ASCII representation of the phonemes.
Documentation from Espressif was somewhat lacking and I had the feeling that this would barely support English, and that if I wanted to use any other language (French in my case), I would have to use a strong English accent to pronounce the words.

Do you have more details on this whole "commands on device" option?

**Comment by nikito** [comment]
From what I understand you just set this in your willow config, and once connected to Home Assistant or OpenHAB it will pull in your entities and create the voice commands:
![image](https://github.com/toverainc/willow/assets/733609/a8c804a4-680b-46fc-96aa-9bff9c7a5329)
![image](https://github.com/toverainc/willow/assets/733609/09e058b4-51fa-4013-a875-8e3b25fd7416)

**Reply by nikito** [reply, marked as answer]
There's also more detailed info in the wiki: https://github.com/toverainc/willow/wiki/Local-Command-Support

**Reply by kristiankielhofner** [reply]
This is generally accurate but I should point out we don't currently support automatic configuration with other endpoints (openHAB, rest) for local voice commands.

Willow (thanks to underlying libraries) is capable of defining these commands dynamically on device. With our work on dynamic configuration users will be able to arbitrarily define commands and automatically (and dynamically) update Willow devices with them and/or update directly from HA based on discovered and selected entities.
